specs/agents/metrics-otel.md

```diff
@@ -0,0 +1,166 @@
+
+## Aggregation
+
+The OpenTelemetry SDK [Aggregations](https://opentelemetry.io/docs/reference/specification/metrics/sdk/#aggregation) define how the observations made on a given Instrument are aggregated into a single value for export. For example, counters are aggregated by summing all values, whereas for gauges the last value is used.
+
+Aggregations are configurable in OpenTelemetry. Currently, this mainly makes sense for Histogram-Instruments. A histogram instrument is usually used when the distribution of the observed values is of interest, e.g. for latency. At the moment the SDK offers [two options](https://opentelemetry.io/docs/reference/specification/metrics/sdk/#histogram-aggregations) for histogram aggregations:
+ * A bucket based histogram with explicitly provided bucket boundaries
+ * An exponential histogram with predefined, exponentially increasing bucket boundaries
+
+In the SDK, the default aggregation to use for each instrument type (counter, gauge, histogram, etc) is defined by the Exporter / MetricsReader. However, users can explicitly override the aggregation per metric by using [Views](https://opentelemetry.io/docs/reference/specification/metrics/sdk/#view). Note that views are an SDK concept and are not part of the OpenTelemetry API.
+
+Agents MUST use the OpenTelemetry SDK [default aggregations per instrument type](https://opentelemetry.io/docs/reference/specification/metrics/sdk/#default-aggregation) as default aggregations for their exporter:
+
+| **Instrument Type**         | **Aggregation**            |
+|-----------------------------|----------------------------|
+|  Counter                    | Sum                        |
+|  Asynchronous Counter       | Sum                        |
+|  UpDownCounter              | Sum                        |
+|  Asynchronous UpDownCounter | Sum                        |
+|  Asynchronous Gauge         | Last Value                 |
+|  Histogram                  | Explicit Bucket Histogram  |
+
+### Histogram Aggregation
+
+The OpenTelemetry SDK uses the following prometheus-inspired bucket boundaries for histograms by default:
+```
+[0, 5, 10, 25, 50, 75, 100, 250, 500, 750, 1000, 2500, 5000, 7500, 10000]
+```
+
+Because IntakeV2 histogram serialization skips empty buckets, we are free to use histogram boundaries which are more accurate and cover a greater range. Therefore, agents SHOULD use the following histogram boundaries by default instead:
+
+```
+[0.00390625, 0.00552427, 0.0078125, 0.0110485, 0.015625, 0.0220971, 0.03125, 0.0441942, 0.0625, 0.0883883, 0.125, 0.176777, 0.25, 0.353553, 0.5, 0.707107, 1, 1.41421, 2, 2.82843, 4, 5.65685, 8, 11.3137, 16, 22.6274, 32, 45.2548, 64, 90.5097, 128, 181.019, 256, 362.039, 512, 724.077, 1024, 1448.15, 2048, 2896.31, 4096, 5792.62, 8192, 11585.2, 16384, 23170.5, 32768, 46341.0, 65536, 92681.9, 131072]
+```
+
+These boundaries are an exponential scale base `sqrt(2)` rounded to six significant figures.
```

Comment:
The suggested boundaries are just what seemed reasonable according to my gut feeling. I had latencies in either seconds or milliseconds as usecase in mind. Please feel free to suggest something better here, I just felt like that the default boundaries from Otel are not good enough.

Reply:
As we're looking at this for the Go Agent (https://github.com/elastic/apm-agent-go/pull/1407), we wonder if there should be boundaries set at all.
Boundaries set at the exporter level will only be overridable if the customer overrides them within that exporter too. If they already provided their own boundaries, we will not keep those, and use our own values.
If the customer is migrating from another exporter to our agent, both exporters will also have different boundaries.

This seems like it can cause confusion to customers, who would expect their boundaries to remain the same, but see something different.

cc @axw

Reply:
> Boundaries set at the exporter level will only be overridable if the customer overrides them within that exporter too

Is this something special about the Go Otel SDK? In Java the aggregations defined on the exporter act as a default if no aggregation was provided for the metric by the user. To my knowledge, the only way for users to customize aggregations is by using [SDK Views](https://opentelemetry.io/docs/reference/specification/metrics/sdk/#view).

We even have a test ensuring that user aggregations provided via views take precedence: [JUnit Test](https://github.com/elastic/apm-agent-java/blob/main/apm-agent-plugins/apm-opentelemetry/apm-opentelemetry-metricsdk-plugin/src/test/java/co/elastic/apm/agent/otelmetricsdk/PrivateUserSdkOtelMetricsTest.java#L99).

> If they already provided their own boundaries, we will not keep those, and use our own values.
Could you elaborate, what it looks like when a user provides his own boundaries in Go?

Reply:
Sorry, I may have been unclear. The Go agent does have the same behavior as you describe for the Java one.
My point is that when using two exporters (say ours, and prometheus), the boundaries will be different, which, in case they migrate from one to the other, would be very confusing to them.

> Could you elaborate, what it looks like when a user provides his own boundaries in Go?

That would be if they provide their own view. Something like this:

```
provider := metric.NewMeterProvider(
	metric.WithReader(ourAgentExporter),
	metric.WithView(metric.NewView(
		metric.Stream{Aggregation: aggregation.ExplicitBucketHistogram{
			Boundaries: []float64{0, 5, 10, 25, 50, 75, 100, 250, 500, 1000},
		}},
	)),
)
```

In this case, despite the customer specifying boundaries themselves, they won't be used, which can lead to more confusion.

Reply:
> In this case, despite the customer specifying boundaries themselves, they won't be used, which can lead to more confusion.

This is different from Java: If the user defines an aggregation for a metric via a View, this aggregation will be used instead of the default aggregation provided with our exporter. This behaviour is verified with the linked JUnit test from my previous post.

I looked at the [Otel spec for views](https://opentelemetry.io/docs/reference/specification/metrics/sdk/#view) and noticed two things regarding the go snippet:
1.  According to the spec the instrument name is a non-optional input when creating a view, I however do not see it in your code. Does the Go implementation deviate there from the spec? In java I had to provide an `InstrumentSelector`:
 ```
 builder.registerView(
            InstrumentSelector.builder().setName("custom_histo").build(),
            View.builder().setAggregation(Aggregation.explicitBucketHistogram(List.of(1.0, 5.0))).build()
        )
 ```
 2. The spec says the following about aggregations defined via Views:
 > The aggregation (optional) to be used. If not provided, the SDK MUST apply a [default aggregation](https://opentelemetry.io/docs/reference/specification/metrics/sdk/#default-aggregation) configurable on the basis of instrument kind according to the [MetricReader](https://opentelemetry.io/docs/reference/specification/metrics/sdk/#metricreader) instance.
 
 This statement says that if not overridden via a view, the default aggregation from the exporter shall be used.
 While not explicitly stated, I would read from that that the following statement is true as well:
 If an aggregation is provided via a View, it shall be used instead of the aggregation from the exporter. Without this, the spec would not make sense as aggregations from views would effectively never be used, because exporters always provide a default.
 
 Does go somehow deviate from the spec here?

Reply:
> According to the spec the instrument name is a non-optional input when creating a view, I however do not see it in your code

Yes, the code above was a simplified example.

And you are right, I missread/interpreted the Go code. If a customer specified a view, that will definitely override our exporter config. 
I added a test for that here: https://github.com/elastic/apm-agent-go/pull/1407/commits/efc516cc861d3eeda5cddafb6a27ae6933323ff5

Reply:
Okay, so I think the only open discussion point is this one:

> My point is that when using two exporters (say ours, and prometheus), the boundaries will be different, which, in case they migrate from one to the other, would be very confusing to them.

My understanding of how users would use the [histogram data type](https://www.elastic.co/guide/en/elasticsearch/reference/current/histogram.html) in elastic is that they would usually not look at the raw data, but apply aggregations, such as percentiles or the `histogram`-aggregation. These aggregations do not rely on the buckets of the raw data. E.g. the `histogram`-aggregation defines the output buckets via the `interval` parameter.

This means that the more accurate the raw data is (=smaller buckets), the better the user experience (e.g. more accurate percentiles). This of course is a tradeoff with the storage consumption.

My (opinionated) viewpoint is that if a user does not provide custom histogram boundaries in Otel via views and currently exports the data to prometheus, he seems to not actually care about the distribution: It is not really possible to compute meaningful percentiles based on the accuracy of the default histogram boundaries. So he most likely wouldn't even notice the usage of different histogram boundaries when switching to the elastic exporter. If he notices, it would be an upgrade in terms of accuracy.

At the same time, we want to make the onboarding experience as easy as possible. As a non-expert user I just want to define a metric and do some meaningful analysis on the data without too much effort.
If we provide a decent histogram aggregation as default, we remove the burden from the user of remembering to also define a view for his metric. Otherwise he will have a bad accuracy and might not even know about it.

Actually @trentm is currently looking into OTel exponential histograms, which have recently been made non-experimental. Those are somewhat similar to HDR-Histograms, which we currently use in the APM server for our transaction metrics. Therefore, we might even think of switching to exponential-histograms as default aggregation for our exporters.

Reply:
I looked a bit today -- but didn't get very far.  I think considering using an exponential histogram aggregation as the default for our exporters would be interesting, and we should consider it before making OTel Metrics "GA" in our agents. I'm curious if OTel SDKs will consider doing so at some point, and if they'd consider that a breaking change.

I think it would take me a while to grok transforming the exponential histogram data points to the histogram format for our intake and I'm keen to get on to 8.8 milestone issues. I think this should be done for follow-up work.

Somes notes:

- Our OTLP intake doesn't yet support exponential histograms. [`apm-data/input/otlp/metrics.go#addMetric()`](https://github.com/elastic/apm-data/blob/3ad1a5caedc9bf5b36ebd62cdc8193e7cedc7d05/input/otlp/metrics.go#L329-L381) would need to add a `case pmetric.MetricTypeExponentialHistogram:` branch ([`MetricType` docs](https://pkg.go.dev/go.opentelemetry.io/collector/pdata/pmetric#MetricType).

- https://gist.github.com/trentm/46ffe98340e5750f043dd489a78d5db2 shows a Node.js script to generate exponential histogram output, including a dump of example `resourceMetrics`.

Reply:
@dmathieu is this discussion resolved for you? If yes, I'm going to go ahead and merge this PR.